File: iOS/GroceryApp/AppConfig.swift
import Foundation

// MARK: - Store Configuration
enum StoreLocation: String, CaseIterable {
    case aa = "aa"
    case nyc = "nyc"
    
    var displayName: String {
        switch self {
        case .aa: return "Ann Arbor Store"
        case .nyc: return "New York City Store"
        }
    }
}

// MARK: - App Configuration
struct AppConfig {

    // MARK: - Current Store Selection
    //
    // Lazily initialized from UserDefaults on first access so cold starts
    // with a persisted session see the correct scope BEFORE DatabaseManager
    // opens the database. Without this, DatabaseManager.init() would spot
    // a mismatch between the default (.nyc) and the store last used by the
    // signed-in user, and spuriously purge the local data.
    //
    // The didSet re-persists to UserDefaults, so callers can simply write
    // `AppConfig.currentStore = .aa` after login and be done.
    private static let persistedStoreKey = "AppConfig.persistedCurrentStore"

    static var currentStore: StoreLocation = {
        if let raw = UserDefaults.standard.string(forKey: persistedStoreKey),
           let loc = StoreLocation(rawValue: raw) {
            return loc
        }
        return .nyc
    }() {
        didSet {
            UserDefaults.standard.set(currentStore.rawValue, forKey: persistedStoreKey)
        }
    }

    /// Maps an authenticated username to its corresponding store.
    /// Credentials in this demo are prefixed with "aa-" or "nyc-";
    /// callers get a single, consistent resolution rule instead of
    /// copying the same `username.contains(...)` check in multiple places.
    static func store(for username: String) -> StoreLocation {
        let lowered = username.lowercased()
        if lowered.contains("aa-store") {
            return .aa
        }
        return .nyc
    }
    
    // MARK: - Capella App Services Configuration (ENV/Info.plist DRIVEN)
    // Prefer environment variables, then Info.plist. Computed each time
    // (not lazy-stored) so an Info.plist populated after the first access
    // during test setup still resolves correctly.
    private static func configValue(for key: String) -> String {
        ProcessInfo.processInfo.environment[key]
            ?? (Bundle.main.object(forInfoDictionaryKey: key) as? String ?? "")
    }

    private static var baseURL: String       { configValue(for: "CBL_BASE_URL") }
    private static var aaDB: String          { configValue(for: "CBL_AA_DB") }
    private static var nycDB: String         { configValue(for: "CBL_NYC_DB") }
    private static var aaUser: String        { configValue(for: "CBL_AA_USER") }
    private static var nycUser: String       { configValue(for: "CBL_NYC_USER") }
    private static var passwordValue: String { configValue(for: "CBL_PASSWORD") }

    // The copilot's App Endpoint. Falls back to the inventory endpoint when unset, so a
    // single-backend setup keeps working with no configuration at all.
    private static var copilotBaseURL: String {
        let value = configValue(for: "CBL_COPILOT_BASE_URL")
        return value.isEmpty ? baseURL : value
    }
    private static var copilotAADB: String {
        let value = configValue(for: "CBL_COPILOT_AA_DB")
        return value.isEmpty ? aaDB : value
    }
    private static var copilotNYCDB: String {
        let value = configValue(for: "CBL_COPILOT_NYC_DB")
        return value.isEmpty ? nycDB : value
    }
    private static var copilotUserValue: String {
        let value = configValue(for: currentStore == .aa
                                ? "CBL_COPILOT_AA_USER" : "CBL_COPILOT_NYC_USER")
        return value.isEmpty ? username : value
    }
    private static var copilotPasswordValue: String {
        let value = configValue(for: "CBL_COPILOT_PASSWORD")
        return value.isEmpty ? passwordValue : value
    }

    static var syncGatewayURL: String {
        switch currentStore {
        case .aa:
            return "\(baseURL)/\(aaDB)"
        case .nyc:
            return "\(baseURL)/\(nycDB)"
        }
    }

    /// Endpoint serving the copilot's collections.
    static var copilotSyncGatewayURL: String {
        switch currentStore {
        case .aa:
            return "\(copilotBaseURL)/\(copilotAADB)"
        case .nyc:
            return "\(copilotBaseURL)/\(copilotNYCDB)"
        }
    }

    static var copilotUsername: String { copilotUserValue }
    static var copilotPassword: String { copilotPasswordValue }

    /// True when the copilot's collections come from a different App Endpoint than
    /// inventory, which means two replicators rather than one.
    static var usesSeparateCopilotEndpoint: Bool {
        copilotSyncGatewayURL != syncGatewayURL
    }

    // MARK: - Which endpoint owns `inventory`
    //
    // `inventory` can only be replicated from ONE endpoint. Two replicators pulling the
    // same collection into the same local collection would each treat the other's writes
    // as remote changes and fight over every document.
    //
    // This matters because Step 1 searches `embedding.text.vector` ON inventory documents:
    //
    //   • `false` — inventory comes from the original endpoint. Safe for the existing app,
    //     but if that endpoint's inventory has no embeddings then semantic search finds
    //     nothing over synced data, and only the bundled seed works.
    //   • `true`  — inventory comes from the copilot endpoint, whose dataset is a superset
    //     (the data-model spec is explicitly additive, so every field the current app reads
    //     is still present). This is what makes Step 1 work against live sync.
    //
    // Set to `true` once the copilot endpoint is confirmed to serve the full inventory
    // superset. Until then the original endpoint keeps ownership and nothing regresses.
    static let copilotEndpointOwnsInventory: Bool = false

    /// Collections replicated from the inventory endpoint.
    static var inventoryEndpointCollections: [String] {
        var collections = [profileCollectionName, ordersCollectionName]
        if !copilotEndpointOwnsInventory { collections.insert(collectionName, at: 0) }
        return collections
    }

    /// Collections replicated from the copilot endpoint. When there is only one endpoint
    /// these are simply added to the single replicator.
    static var copilotEndpointCollections: [String] {
        var collections = [planogramsCollectionName, knowledgeCollectionName,
                           tasksCollectionName]
        if copilotEndpointOwnsInventory { collections.insert(collectionName, at: 0) }
        return collections
    }
    
    static var username: String {
        switch currentStore {
        case .aa:
            return aaUser
        case .nyc:
            return nycUser
        }
    }
    
    static var password: String {
        return passwordValue
    }
    
    static var storeId: String {
        switch currentStore {
        case .aa:
            return "aa-store-01"
        case .nyc:
            return "nyc-store-01"
        }
    }
    
    // MARK: - Database Configuration
    static let databaseName = "GroceryInventoryDB"
    
    // Capella Collection Configuration (matches server-side setup)
    static var scopeName: String {
        switch currentStore {
        case .aa:
            return "AA-Store"
        case .nyc:
            return "NYC-Store"
        }
    }
    
    static let collectionName = "inventory"  // Main inventory collection
    static let ordersCollectionName = "orders"  // Orders collection
    static let profileCollectionName = "profile"  // Store profile collection

    // Collections added by the edge vector search / copilot extension.
    // These sit in the existing per-store scopes and sync under the existing
    // single App User per store — no new roles, users, or channels.
    static let planogramsCollectionName = "planograms"          // Step 2: shelf audit
    static let knowledgeCollectionName = "product_knowledge"    // Step 3: RAG source chunks
    static let tasksCollectionName = "tasks"                    // Request Help

    /// Every collection the replicator should carry, in a single list so the sync
    /// manager and the local seeder cannot drift apart.
    static var allSyncedCollections: [String] {
        [collectionName, profileCollectionName, ordersCollectionName,
         planogramsCollectionName, knowledgeCollectionName, tasksCollectionName]
    }
    
    // MARK: - Sync Configuration (Event-Driven, Real-Time)
    // NOTE: "Continuous" sync is EVENT-DRIVEN, not polling!
    // - Keeps WebSocket connection open to Capella
    // - Changes are pushed/pulled IMMEDIATELY when they occur
    // - No polling or timers - true real-time bidirectional sync
    static let syncHeartbeat: UInt16 = 60 // WebSocket keepalive (not data polling!)
    static let syncMaxAttempts: Int = 10
    static let syncMaxAttemptWaitTime: TimeInterval = 300 // 5 minutes
    static let syncContinuous: Bool = true // TRUE = Event-driven real-time sync
    static let syncAllowBackground: Bool = true
    
    // MARK: - Feature Flags
    static let enableAppServicesSync: Bool = true
    static let enableP2PSync: Bool = true
    static let enableAutoDataSeeding: Bool = false // DISABLED: No more hard-coded data

    // MARK: - Store Associate Copilot (edge vector search)

    /// Seeds the bundled extended dataset into the local database when a collection
    /// comes up empty. This exists so the copilot can be built, demoed and explored
    /// with no Capella backend at all — the extended dataset is not in Capella yet,
    /// and a developer picking this app up should not need a cloud account to see
    /// vector search work. Documents that arrive over App Services supersede the
    /// seeded copies normally, since they carry the same document IDs.
    static let enableLocalDatasetSeeding: Bool = true

    /// Whether footwear leads the *demo script* — suggested queries, the default planogram,
    /// and example copy.
    ///
    /// The narrative is grocery-first, so this is `false`: the suggestions, the shelf the
    /// audit opens on, and the RAG prompts are all grocery. That is a presentation choice,
    /// not a data rule.
    ///
    /// Deliberately does NOT filter the data. Per the Jul 2026 discussion the app is
    /// generic — what it can find is decided by the documents and embeddings present, not by
    /// a hardcoded category list. A real associate searching a Target or Walmart-style store
    /// should be able to find a shoe if the store stocks one. So footwear stays searchable,
    /// retrievable and seeded; it simply is not what the demo leads with.
    ///
    /// Set to `true` to put footwear back in the scripted flow, which demonstrates the
    /// "same capability, new category" generality claim.
    static let footwearNarrativeEnabled: Bool = false

    /// Categories excluded from copilot queries.
    ///
    /// Empty by design — see `footwearNarrativeEnabled`. The hook is kept because the query
    /// builders read it, so a category can be suppressed for a specific demo without
    /// touching the query code, but the default is to hide nothing.
    static var hiddenCategories: [String] { [] }

    /// Cosine-distance ceiling for a result to count as relevant.
    ///
    /// NOT the spec's hardcoded 0.35 — that was written before real vectors existed.
    /// Measured against the actual MiniLM corpus, the hero query's best match sits at
    /// 0.24 and the 5th percentile of all 104 documents is 0.57, so 0.35 would surface
    /// only 2 documents and drop legitimate near-misses. 0.60 keeps the relevant set
    /// and still excludes the bulk of the catalogue (median distance 0.81).
    /// Tunable at runtime from the copilot's behind-the-scenes screen.
    static let defaultRelevanceThreshold: Double = 0.60

    /// How many candidates a vector query returns before threshold filtering.
    static let copilotSearchLimit: Int = 10

    /// Chunks fed to the on-device SLM for RAG answers.
    static let copilotRAGChunkCount: Int = 4
    
    // MARK: - Debug Configuration
    static let debugLogging: Bool = true
    
    // MARK: - Helper Methods
    
    static func printConfiguration() {
        print("📋 ========================================")
        print("📋 App Configuration")
        print("📋 ========================================")
        print("📋 Store: \(currentStore.displayName)")
        print("📋 Sync URL: \(syncGatewayURL)")
        print("📋 Username: \(username)")
        print("📋 Store ID: \(storeId)")
        print("📋 Database: \(databaseName)")
        print("📋 Scope: \(scopeName)")
        print("📋 Collections: \(collectionName), \(ordersCollectionName), \(profileCollectionName)")
        print("📋 App Services Sync: \(enableAppServicesSync ? "✅ Enabled" : "❌ Disabled")")
        print("📋 P2P Sync: \(enableP2PSync ? "✅ Enabled" : "❌ Disabled")")
        print("📋 Auto Data Seeding: \(enableAutoDataSeeding ? "✅ Enabled" : "❌ Disabled")")
        print("📋 ========================================")
    }
    
    static func getEnvironmentInfo() -> [String: Any] {
        return [
            "store": currentStore.rawValue,
            "store_name": currentStore.displayName,
            "sync_url": syncGatewayURL,
            "username": username,
            "store_id": storeId,
            "database": databaseName,
            "collection": collectionName,
            "app_services_enabled": enableAppServicesSync,
            "p2p_enabled": enableP2PSync,
            "auto_seeding_enabled": enableAutoDataSeeding
        ]
    }
}

File: Android/app/src/main/java/com/example/groceryapplication/AppConfig.kt
package com.example.groceryapplication

import android.util.Log
import com.example.groceryapplication.BuildConfig

/**
 * Store Configuration
 */
enum class StoreLocation(val displayName: String) {
    AA("Ann Arbor Store"),
    NYC("New York City Store")
}

/**
 * Centralized App Configuration
 * Matches iOS AppConfig for consistency
 */
object AppConfig {
    
    // MARK: - Current Store Selection
    // Set dynamically at login based on user credentials
    var currentStore: StoreLocation = StoreLocation.NYC
        private set
    
    fun setStoreForUser(username: String) {
        currentStore = when {
            username.lowercase().startsWith("aa-") -> StoreLocation.AA
            username.lowercase().startsWith("nyc-") -> StoreLocation.NYC
            else -> {
                Log.w("AppConfig", "⚠️ Unknown user prefix for: $username, defaulting to NYC")
                StoreLocation.NYC
            }
        }
        Log.d("AppConfig", "🏪 Store set to ${currentStore.displayName} for user: $username")
    }
    
    // MARK: - Capella App Services Configuration (ENV-DRIVEN)
    // Values are injected via BuildConfig (from Gradle properties or env vars)
    private val BASE_URL: String = BuildConfig.CBL_BASE_URL
    private val AA_DB: String = BuildConfig.CBL_AA_DB
    private val NYC_DB: String = BuildConfig.CBL_NYC_DB
    private val AA_USER: String = BuildConfig.CBL_AA_USER
    private val NYC_USER: String = BuildConfig.CBL_NYC_USER
    private val PASSWORD: String = BuildConfig.CBL_PASSWORD
    
    val syncGatewayURL: String
        get() = when (currentStore) {
            StoreLocation.AA -> "$BASE_URL/$AA_DB"
            StoreLocation.NYC -> "$BASE_URL/$NYC_DB"
        }
    
    val username: String
        get() = when (currentStore) {
            StoreLocation.AA -> AA_USER
            StoreLocation.NYC -> NYC_USER
        }
    
    val password: String
        get() = PASSWORD
    
    val storeId: String
        get() = when (currentStore) {
            StoreLocation.AA -> "aa-store-01"
            StoreLocation.NYC -> "nyc-store-01"
        }
    
    // MARK: - Database Configuration
    const val DATABASE_NAME = "GroceryInventoryDB"
    
    // Capella Collection Configuration (matches server-side setup)
    val scopeName: String
        get() = when (currentStore) {
            StoreLocation.AA -> "AA-Store"
            StoreLocation.NYC -> "NYC-Store"
        }
    
    const val COLLECTION_NAME = "inventory"  // Main inventory collection
    const val ORDERS_COLLECTION_NAME = "orders"  // Orders collection
    const val PROFILE_COLLECTION_NAME = "profile"  // Store profile collection

    // Collections added by the edge vector search / copilot extension. These sit in the
    // existing per-store scopes and sync under the existing single App User per store —
    // additive channel configuration, not an access-model change.
    const val PLANOGRAMS_COLLECTION_NAME = "planograms"          // Step 2: shelf audit
    const val KNOWLEDGE_COLLECTION_NAME = "product_knowledge"    // Step 3: RAG source chunks
    const val TASKS_COLLECTION_NAME = "tasks"                    // Request Help

    /**
     * Every collection the replicator should carry, in one list so the sync manager and the
     * local seeder cannot drift apart.
     */
    val allSyncedCollections: List<String>
        get() = listOf(
            COLLECTION_NAME, PROFILE_COLLECTION_NAME, ORDERS_COLLECTION_NAME,
            PLANOGRAMS_COLLECTION_NAME, KNOWLEDGE_COLLECTION_NAME, TASKS_COLLECTION_NAME
        )

    // MARK: - Store Associate Copilot (edge vector search)

    /**
     * Seeds the bundled extended dataset into the local database when a collection comes up
     * empty.
     *
     * **Off, because Capella is now the source of truth.** The App Endpoint serves all six
     * collections with real MiniLM embeddings, so the app pulls everything over the
     * replicator. Leaving this on was not wrong — synced documents carry the same IDs and
     * supersede seeded ones — but it wrote a redundant revision to every document and made it
     * impossible to tell from the UI whether sync was actually working.
     *
     * Set to `true` to work with no backend at all: useful offline, and a reasonable fallback
     * if the free-tier cluster has hibernated, since a hibernated backend with seeding off
     * means an empty inventory rather than a stale one.
     */
    const val ENABLE_LOCAL_DATASET_SEEDING = false

    /**
     * Cosine-distance ceiling for a result to count as relevant.
     *
     * Not the data-model spec's hardcoded 0.35 — that predates real vectors. Measured against
     * the actual MiniLM corpus, the hero query's best match sits at 0.24 and the 5th
     * percentile of all documents at 0.57, so 0.35 would surface almost nothing. Matches the
     * iOS default; note the bundled int8 model reads roughly 0.02 higher than iOS's fp16
     * CoreML model for the same query, which this threshold comfortably absorbs.
     */
    const val DEFAULT_RELEVANCE_THRESHOLD = 0.60

    /** How many candidates a vector query returns before threshold filtering. */
    const val COPILOT_SEARCH_LIMIT = 10

    /** Chunks retrieved for a grounded answer. */
    const val COPILOT_RAG_CHUNK_COUNT = 4

    /**
     * Whether footwear leads the demo script — suggested queries and example copy.
     *
     * The narrative is grocery-first, so this is false. It is a presentation choice, not a
     * data rule: footwear stays searchable and seeded, because what the app can find is
     * decided by the documents and embeddings present, not by a hardcoded category list.
     */
    const val FOOTWEAR_NARRATIVE_ENABLED = false
    
    // MARK: - Sync Configuration (Event-Driven, Real-Time)
    // NOTE: "Continuous" sync is EVENT-DRIVEN, not polling!
    // - Keeps WebSocket connection open to Capella
    // - Changes are pushed/pulled IMMEDIATELY when they occur
    // - No polling or timers - true real-time bidirectional sync
    const val SYNC_HEARTBEAT: Long = 60 // WebSocket keepalive (not data polling!)
    const val SYNC_MAX_ATTEMPTS = 10
    const val SYNC_MAX_ATTEMPT_WAIT_TIME: Long = 300 // 5 minutes
    const val SYNC_CONTINUOUS = true // TRUE = Event-driven real-time sync
    const val SYNC_ALLOW_BACKGROUND = true
    
    // MARK: - Feature Flags
    const val ENABLE_APP_SERVICES_SYNC = true
    const val ENABLE_P2P_SYNC = true
    const val ENABLE_AUTO_DATA_SEEDING = false // DISABLED: No more hard-coded data
    
    // MARK: - P2P (Peer-to-Peer) Configuration
    // MultipeerReplicator settings for local device-to-device sync
    // NOTE: This MUST match iOS peer group ID for cross-platform sync!
    const val P2P_PEER_GROUP_ID = "com.example.groceryapp"  // Network identifier for discovering peers (matches iOS)
    const val P2P_IDENTITY_LABEL = "com.example.groceryapp.p2p.identity"  // TLS identity label
    const val P2P_AUTO_START = true  // Auto-start P2P sync on app launch
    
    // MARK: - Debug Configuration
    const val DEBUG_LOGGING = true
    
    // MARK: - Helper Methods
    
    fun printConfiguration() {
        Log.d("AppConfig", "📋 ========================================")
        Log.d("AppConfig", "📋 App Configuration")
        Log.d("AppConfig", "📋 ========================================")
        Log.d("AppConfig", "📋 Store: ${currentStore.displayName}")
        Log.d("AppConfig", "📋 Sync URL: $syncGatewayURL")
        Log.d("AppConfig", "📋 Username: $username")
        Log.d("AppConfig", "📋 Store ID: $storeId")
        Log.d("AppConfig", "📋 Database: $DATABASE_NAME")
        Log.d("AppConfig", "📋 Scope: $scopeName")
        Log.d("AppConfig", "📋 Collections: $COLLECTION_NAME, $ORDERS_COLLECTION_NAME, $PROFILE_COLLECTION_NAME")
        Log.d("AppConfig", "📋 App Services Sync: ${if (ENABLE_APP_SERVICES_SYNC) "✅ Enabled" else "❌ Disabled"}")
        Log.d("AppConfig", "📋 P2P Sync: ${if (ENABLE_P2P_SYNC) "✅ Enabled" else "❌ Disabled"}")
        Log.d("AppConfig", "📋 Auto Data Seeding: ${if (ENABLE_AUTO_DATA_SEEDING) "✅ Enabled" else "❌ Disabled"}")
        Log.d("AppConfig", "📋 ========================================")
    }
    
    fun getEnvironmentInfo(): Map<String, Any> {
        return mapOf(
            "store" to currentStore.name,
            "store_name" to currentStore.displayName,
            "sync_url" to syncGatewayURL,
            "username" to username,
            "store_id" to storeId,
            "database" to DATABASE_NAME,
            "scope" to scopeName,
            "collection" to COLLECTION_NAME,
            "app_services_enabled" to ENABLE_APP_SERVICES_SYNC,
            "p2p_enabled" to ENABLE_P2P_SYNC,
            "auto_seeding_enabled" to ENABLE_AUTO_DATA_SEEDING
        )
    }
}

